Addition of a DialoguePipeline

guillaume-be opened the pull request:
- Addition of a Conversation object to keep track of multi-turn conversation
- Creation of a DialoguePipeline to process Conversations using the history context
- Integration tests for DialoguePipeline using `microsoft/DialoGPT-medium`

guillaume-be commented:
This is a back-port of https://github.com/guillaume-be/rust-bert/pull/57. I did not implement the ConversationManager as I felt it did not quite fit the general API of this library. I however added the concept of `Conversations` keeping track of past user inputs, generated responses and the history token ids. Conversations include a print option that will display the entire dialogue.

```python
print(conversation)
```
```
Conversation id: 2716da3e-8cde-4071-97bc-218d88764b7b 
user >> What's the last book you have read? 
bot >> The Last Question 
user >> Why do you recommend it? 
bot >> It's a good book. 
```

patrickvonplaten commented on src/transformers/pipelines.py:
```diff
@@ -1664,6 +1665,270 @@ def __call__(
+        conversation_1 = Conversation("Going to the movies tonight - any suggestions?")
+        conversation_2 = Conversation("What's the last book you have read?")
+
+        conversation_pipeline([conversation_1, conversation_2])
```
```suggestion
        dialogue_pipeline([conversation_1, conversation_2])
```

patrickvonplaten commented on src/transformers/pipelines.py:
```diff
@@ -1664,6 +1665,270 @@ def __call__(
+
+        conversation_1.add_user_input("Is it an action movie?")
+
+        conversation_pipeline([conversation_1, conversation_2])
```
```suggestion
        dialogue_pipeline([conversation_1, conversation_2])
```

patrickvonplaten commented on src/transformers/pipelines.py:
```diff
@@ -1664,6 +1665,270 @@ def __call__(
+            self.pad_token_id = self.tokenizer.eos_token_id
+        self.min_response_allowed_length = kwargs.get("min_response_allowed_length", 32)
+
+    def __call__(self, *args, clean_up_tokenization_spaces=True, **generate_kwargs):
```
Can we call `args` => `conversations`?

patrickvonplaten commented on src/transformers/pipelines.py:
```diff
@@ -1664,6 +1665,270 @@ def __call__(
+                    [conversation.new_user_input for conversation in active_conversations]
+                )
+                histories = [conversation.history for conversation in active_conversations]
+                max_length = generate_kwargs.get("max_length", 1000)
```
I think I would prefer to not set the default `max_length` to 1000 here. The user can set the default value for each model individually in the model's config (under task specific params) => compare for example with XLNet here: https://s3.amazonaws.com/models.huggingface.co/bert/xlnet-base-cased-config.json .

patrickvonplaten replied:
```suggestion
                max_length = generate_kwargs.get("max_length", self.model.config.max_length)
```

guillaume-be replied:
I agree the value of 1000 is arbitrary (taken from the illustrative example in the model card). The issue is that the `DialoGPT` configuration files do not set a `max_length`. To my understanding, this means that without specifying it to the `generate` method, it will be set to the GPT2 default, that is 20. This seems very low for a conversation pipeline as the input alone is likely to exceed this value. I am not sure defaulting to the configuration value is going to be a good user experience. Maybe the way to go is to force the user to provide a value? Or maybe update the configuration of the `DialoGPT` configuration files?

patrickvonplaten commented on src/transformers/pipelines.py:
```diff
@@ -1664,6 +1665,270 @@ def __call__(
+                            input_length, max_length
+                        )
+                    )
+                generate_kwargs["max_length"] = max_length
```
```suggestion
```

patrickvonplaten commented on src/transformers/pipelines.py:
```diff
@@ -1664,6 +1665,270 @@ def __call__(
+
+                cleaned_history = self._clean_padding_history(generated_responses)
+                if isinstance(args[0], Conversation):
+                    args[0].mark_processed()
```
would be nice to rename `args` to conversations

patrickvonplaten commented on src/transformers/pipelines.py:
```diff
@@ -1664,6 +1665,270 @@ def __call__(
+        Builds an input prepended by the history for this conversation, allowing multi-turn conversation with context
+        """
+        outputs = []
+        for input, history in zip(inputs, histories):
```
Very nice! I like it

patrickvonplaten commented on src/transformers/pipelines.py:
```diff
@@ -1664,6 +1665,270 @@ def __call__(
+            on the associated CUDA device id.
+    """
+
+    def __init__(self, *args, **kwargs):
```
I think as it is implemented now `dialogue_pipeline = pipeline("dialogue", min_response_allowed_length=32)` would throw an error because it is passed to `super().__init__(*args, **kwargs)` => can we just change it to:
```python
def __init__(self, min_respones_allowed_length=32, *args, **kwargs):
   super().__init__(*args, **kwargs)
```

patrickvonplaten replied:
and maybe the name `min_length_for_response` is better here. The word `allowed` is confusing me a bit. What do you think @guillaume-be ?

guillaume-be replied:
Good catch and agreed - will update

patrickvonplaten commented on tests/test_pipelines.py:
```diff
@@ -341,10 +343,74 @@ def test_tf_text_generation(self):
+    @require_torch
+    def test_integration_torch_dialogue(self):
+        # When
+        nlp = pipeline(task="dialogue", device=DEFAULT_DEVICE_NUM)
```
Maybe pass `min_response_allowed_length` or (IMO clearer name `min_length_for_response`) here to test it.

guillaume-be commented:
@patrickvonplaten Thank you so much for the detailed review! I have pushed some changes addressing most of the suggested changes.

A few points still open for discussion:
- I am still unsure about using the default `max_length` from the model_config, especially since the default model used for the pipeline `DialoGPT` does not give this value and it would default to a very low value.
- I have updated the history truncation to take out entire spans (delimited by EOS tokens). I believe this would lead to a more natural truncation of a conversation (skip the first N turn) rather than possibly cutting an input/response in the middle
- You are right that the API differs a bit from the other pipeline (although the QA pipeline also takes special inputs). Here the challenge is that the conversations are by definition stateful. I actually started the Rust implementation storing the conversations in the model itself, allowing simpler inputs to be passed on. This requires the model to be mutable, which can be problematic for multithreading. Now in Python you may not have this issue, but you may want to deploy an conversational application that scales and deploys multiple conversation pipeline workers / containers (the Rust compiler actually points out to an interesting design decision). If you make these stateful, you'll have to ensure that you always send the conversation to the correct worker which can be sub-optimal for load balancing. I believe carrying the history in the conversation itself would be easier to handle for the full system that may leverage this pipeline. Happy to change if you'd like to allow the user to pass a simpler input, but a conversation `id` would probably be required anyway. In any case happy to expose `Conversation` to the library API if the design remains as is.

Thanks again for the feedback - looking forward to your thoughts on these few points.

patrickvonplaten commented:
Thanks a lot for the detailed answer @guillaume-be,

1. (answering to your comment above) - exactly, I would update the config on AWS 
    and it would then look like this:
```
{
  "activation_function": "gelu_new",
  "architectures": [
    "GPT2LMHeadModel"
  ],
  "attn_pdrop": 0.1,
  "bos_token_id": 50256,
  "embd_pdrop": 0.1,
  "eos_token_id": 50256,
  "initializer_range": 0.02,
  "layer_norm_epsilon": 1e-05,
  "model_type": "gpt2",
  "n_ctx": 1024,
  "n_embd": 1024,
  "n_head": 16,
  "n_layer": 24,
  "n_positions": 1024,
  "resid_pdrop": 0.1,
  "summary_activation": null,
  "summary_first_dropout": 0.1,
  "summary_proj_to_labels": true,
  "summary_type": "cls_index",
  "summary_use_proj": true,
  "vocab_size": 50257,
  "task_specific_params" : {
      "dialogue": { 
          "max_length": 1000
      }
  }
}
```
=> This way in the `__init__` of `Pipeline` the `task_specific_params` will overwrite the default params and the `max_length` will be 1000. I think this has following advantage: If the user wants to define a specific `max_length` via the model config, he could update or create a config to `model.config.task_specific_params.dialogue.max_length = 1000` and does not have to input any generate kwargs to the `__call__` function (otherwise the config would always be overwritten by 1000). As soon as we merge this PR, I can update all the DialoGPT configs with the `task_specific_params`.

2. Very nice! This indeed makes more sense and should lead to good (possible never-ending chats) :-) 

3. I agree - I think I'm also in favor of your design choice here. Sadly, I don't have the necessary Rust knowledge to have a better understanding of the consequences of stateful vs. stateless, but it makes a lot of sense what you say! Maybe looping in @mfuntowicz  here to see what he thinks :-)

julien-c commented on src/transformers/pipelines.py:
```diff
@@ -1738,6 +2003,12 @@ def __call__(
         "pt": AutoModelWithLMHead if is_torch_available() else None,
         "default": {"model": {"pt": "gpt2", "tf": "gpt2"}},
     },
+    "dialogue": {
```
can we call this "conversational" to be compatible out of the box with the models that are already tagged on https://huggingface.co/models?filter=conversational ?

I think dialogue might be the better term for the NLP research community but conversational is better for ML practitioners (who are the target audience of pipelines)

julien-c commented on src/transformers/pipelines.py:
```diff
@@ -1664,6 +1665,270 @@ def __call__(
+        return output
+
+
+class DialoguePipeline(Pipeline):
```
`ConversationalPipeline`?

julien-c commented on src/transformers/pipelines.py:
```diff
@@ -1664,6 +1665,270 @@ def __call__(
             return results
 
 
+class Conversation:
```
This abstraction seems a bit weird to me because we are mutating so much of its properties from the pipeline below. i.e. it's not very self-contained. I'm wondering if we should move some of the mutating methods to the class directly.

guillaume-be replied:
I created some class functions to avoid direct field mutation - hopefully it goes in the right direction

julien-c commented on src/transformers/pipelines.py:
```diff
@@ -1664,6 +1665,270 @@ def __call__(
+            self.pad_token_id = self.tokenizer.eos_token_id
+        self.min_length_for_response = min_length_for_response
+
+    def __call__(self, conversations, clean_up_tokenization_spaces=True, **generate_kwargs):
```
I'd be in favor of typing `conversations` here (`Union[Conversation, List[Conversation]]`)

julien-c commented on src/transformers/pipelines.py:
```diff
@@ -1664,6 +1665,270 @@ def __call__(
+                assert isinstance(
+                    conversation, Conversation
+                ), "DialoguePipeline expects a Conversation or list of Conversations as an input"
+                if conversation.new_user_input is None:
```
Do we really need the concept of active conversation? Or would it be handled in the application before entering the pipeline anyways?

It seems to be making the code kinda longer and complex (with active_indexes, etc.)

guillaume-be replied:
I like the concept of `new_user_input` because it plays well with the history kept in the conversation itself. Without the `new_user_input`, the part of the user inputs that need to be tokenized before constructing the model input may be ambiguous. 

Because of this I believe the pipeline needs to be able to handle any conversation passed to it. Pipelines without user input should not be passed through the model (and therefore will not participate to the batch creation). The model outputs then need to be reallocated to the right conversation, requiring the indices.

Would you prefer the application to raise a ValueError if an inactive conversation has been passed to the pipeline? This would  indeed simplify the code, but be less forgiving to potential users. I am open to both - or maybe you had another idea in mind?

julien-c replied:
Personally I would raise a `ValueError` indeed in case of inactive conversation, but up to you:)

guillaume-be replied:
Makes sense (I took this from the Rust implementation that was using a ConversationManager instead of single Conversations passed to it - where inactive conversations have to be dealth with). In this case I agree the responsibility could be passed on to the user. Will update and simplify accordingly

julien-c commented on src/transformers/pipelines.py:
```diff
@@ -1664,6 +1665,270 @@ def __call__(
+                        )
+                    )
+                    conversations.history = cleaned_history[0]
+                    output = conversations
```
Just `return conversations` here?

julien-c commented on src/transformers/pipelines.py:
```diff
@@ -1664,6 +1665,270 @@ def __call__(
+        outputs = [output + [self.pad_token_id] * (max_len - len(output)) for output in outputs]
+        outputs = self.tokenizer.batch_encode_plus(
+            outputs, add_special_tokens=False, is_pretokenized=True, return_tensors=self.framework, padding=False
+        )
```
If I understand correctly outputs is already list of list of ints before this line so why do we pass it to `self.tokenizer`? Is it because it's the simplest way to tensorize to pytorch vs tf? (if it is, the intent is not super explicit)

guillaume-be replied:
Yes - this was the intent. I didn't find a utility class in the library that would conveniently convert a list of integers to a Torch or Tensorflow tensor the way this class did. Is there a better way?

julien-c replied:
Can you create a helper to do this in this file?

guillaume-be replied:
I actually use the `BatchEncoding` directly from the generated `input_ids`. Cleans up quite a bit and better represents the intent, thank you for the suggestion

julien-c commented on src/transformers/pipelines.py:
```diff
@@ -1664,6 +1665,270 @@ def __call__(
+                    if cutoff_eos_index == 0 or cutoff_eos_index == len(input):
+                        break
+                    else:
+                        input = input[cutoff_eos_index + 1 :]
```
I think it's great that you cut off at an `eos`

julien-c commented on tests/test_pipelines.py:
```diff
@@ -341,10 +343,100 @@ def test_tf_text_generation(self):
+    @require_torch
+    def test_integration_torch_dialogue(self):
+        # When
+        nlp = pipeline(task="dialogue", device=DEFAULT_DEVICE_NUM)
```
```suggestion
        nlp = pipeline(task="conversational", device=DEFAULT_DEVICE_NUM)
```

julien-c commented:
Thanks @guillaume-be for your PR!

I left a bunch of comments, but this is in good shape, can't wait to use it on https://huggingface.co/microsoft/DialoGPT-large 😄 

Re. statefulness, yes, we definitely want to keep a stateless system.

As an example, the way our https://convai.huggingface.co/ demo from last year works (and probably the way our conversational widget powered by this PR will work too 😉) is, we store an array of `Message` on the client side, grow it, and send it with each request. It's text so the bandwidth is not an issue.

Message looks like:

```
interface Message {
	incoming: boolean; // <- is it from bot or user
	content:  string;
}
```

julien-c commented:
LGTM

guillaume-be commented:
@julien-c @patrickvonplaten I believe all comments have been addressed - please let me know if I have missed anything. Just resolved the conflict with master

julien-c commented:
Given that `patrickvonplaten` is off for one more week I believe, do you want to give this PR a last look-over @sgugger and merge if it's fine?

LysandreJik commented on src/transformers/pipelines.py:
```diff
@@ -1813,6 +2062,12 @@ def __call__(
     },
+    "conversational": {
+        "impl": ConversationalPipeline,
+        "tf": TFAutoModelWithLMHead if is_tf_available() else None,
+        "pt": AutoModelWithLMHead if is_torch_available() else None,
```
```suggestion
        "tf": TFAutoModelForCausalLM if is_tf_available() else None,
        "pt": AutoModelForCausalLM if is_torch_available() else None,
```

LysandreJik commented on src/transformers/pipelines.py:
```diff
@@ -1739,6 +1741,253 @@ def __call__(
+    Multi-turn conversational pipeline.
+
+    Usage::
+        conversational_pipeline = pipeline("conversational")
+
+        conversation_1 = Conversation("Going to the movies tonight - any suggestions?")
+        conversation_2 = Conversation("What's the last book you have read?")
+
+        conversational_pipeline([conversation_1, conversation_2])
+
+        conversation_1.add_user_input("Is it an action movie?")
+
+        conversational_pipeline([conversation_1, conversation_2])
```
This example does not work for me. It fails with the following:

```
ValueError: Conversation with UUID <class 'uuid.UUID'> does not contain new user input to process. Add user inputs with the conversation's `add_user_input` method
```

A user input must be added to the second conversation as well for that example to work.
```suggestion
        conversational_pipeline = pipeline("conversational")

        conversation_1 = Conversation("Going to the movies tonight - any suggestions?")
        conversation_2 = Conversation("What's the last book you have read?")

        conversational_pipeline([conversation_1, conversation_2])

        conversation_1.add_user_input("Is it an action movie?")
        conversation_2.add_user_input("What is the genre of this book?")

        conversational_pipeline([conversation_1, conversation_2])
```

LysandreJik commented on src/transformers/pipelines.py:
```diff
@@ -1739,6 +1741,253 @@ def __call__(
+    """
+    Multi-turn conversational pipeline.
+
+    Usage::
+        conversational_pipeline = pipeline("conversational")
```
Needs a line return to be correctly rendered in the docs
```suggestion
    Usage::
    
        conversational_pipeline = pipeline("conversational")
```

LysandreJik commented on src/transformers/pipelines.py:
```diff
@@ -1739,6 +1741,253 @@ def __call__(
+        Builds an input prepended by the history for this conversation, allowing multi-turn conversation with context
+        """
+        outputs = []
+        for input, history in zip(inputs, histories):
```
Would it be possible to change `input` to something not shadowing the built-in `input`?

LysandreJik commented on src/transformers/pipelines.py:
```diff
@@ -1739,6 +1741,253 @@ def __call__(
+            if len(input) > max_length - self.min_length_for_response:
+                cutoff_eos_index = 0
+                while len(input) - cutoff_eos_index > max_length - self.min_length_for_response:
+                    cutoff_eos_index = input[cutoff_eos_index:].index(self.tokenizer.eos_token_id)
+                    if cutoff_eos_index == 0 or cutoff_eos_index == len(input) - 1:
+                        break
+                    else:
+                        input = input[cutoff_eos_index + 1 :]
```
This should also break when the cutoff_eos_index is larger than the length of the remaining input. Otherwise it fails because `input[cutoff_eos_index:]` returns an empty list, which cannot be indexed with `self.tokenizer.eos_token_id`.

An easy fix is the following, which could probably be made cleaner.

```suggestion
                    if cutoff_eos_index >= len(input):
                        break
                        
                    cutoff_eos_index = input[cutoff_eos_index:].index(self.tokenizer.eos_token_id)
                    if cutoff_eos_index == 0 or cutoff_eos_index == len(input) - 1:
                        break
                    else:
                        input = input[cutoff_eos_index + 1 :]
```

guillaume-be commented:
@LysandreJik Thank you very much for the review. Good catch on the behaviour of the `eos` token cut-off. I have updated based on  your suggestions, and added docstrings to the `Conversation` class. I have also added both `Conversation` and `ConversationalPipeline` to the top-level `__init__` for consistency with the other pipelines.

sgugger commented on src/transformers/pipelines.py:
```diff
@@ -1895,6 +1898,321 @@ def __call__(
+    ):
+        r"""
+        Args:
+            *conversations: (list of `:class:Conversation`) Conversations to generate responses for
```
```suggestion
            conversations: (list of :class:`~transformers.pipelines.Conversation`) Conversations to generate responses for
```

sgugger commented:
Thanks for the PR @guillaume-be 
Docstrings could be improved but I'll clean up the docs in the pipelines file soon, so will take of that. For future PRs, please remember that `thing` will render thing in italics in the docs, and not in code (you have to use ``thing`` or :obj:`thing`).

guillaume-be commented:
@sgugger Thank you for the review - was indeed a typo on my end. The tests got triggered again and unfortunately a hash verification on torch fails. Coul dyou please restart the build if you have a chance?

julien-c commented:
This is awesome, congrats everyone on shipping this! 🔥

sshleifer commented:
`test_torch_conversation` and `test_integration_torch_conversation` are broken on github actions CI. Could someone fix or delete? https://github.com/huggingface/transformers/runs/1289790225?check_suite_focus=true

LysandreJik commented:
Will take a look.

thomwolf commented:
Should be fixed in https://github.com/huggingface/transformers/pull/7970
